Add PAG support to StableDiffusionControlNetPAGInpaintPipeline

juancopi81 opened the pull request:
# What does this PR do?
Adds PAG (Perturbed-Attention Guidance) support for SD 1.5 with controlnet and inpainting models (StableDiffusionControlNetPAGInpaintPipeline )
Continuation of https://github.com/huggingface/diffusers/issues/8710. It was not mentioned on #8710 but I think this pipeline is pretty cool also 😄 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?
@a-r-r-o-w @yiyixuxu
Anyone in the community is free to review the PR once the tests have passed.

Code example
```py
import cv2
from diffusers import AutoPipelineForInpainting, ControlNetModel, DDIMScheduler
from diffusers.utils import load_image
import numpy as np
from PIL import Image
import torch

init_image = load_image("https://huggingface.co/datasets/diffusers/test-arrays/resolve/main/stable_diffusion_inpaint/boy.png")
init_image = init_image.resize((512, 512))

mask_image = load_image("https://huggingface.co/datasets/diffusers/test-arrays/resolve/main/stable_diffusion_inpaint/boy_mask.png")
mask_image = mask_image.resize((512, 512))

def make_canny_condition(image):
    image = np.array(image)
    image = cv2.Canny(image, 100, 200)
    image = image[:, :, None]
    image = np.concatenate([image, image, image], axis=2)
    image = Image.fromarray(image)
    return image

control_image = make_canny_condition(init_image)

controlnet = ControlNetModel.from_pretrained("lllyasviel/control_v11p_sd15_inpaint", torch_dtype=torch.float16)

pipe = AutoPipelineForInpainting.from_pretrained(
    "jayparmr/icbinp_v8_inpaint_v2", controlnet=controlnet, torch_dtype=torch.float16, enable_pag=True, requires_safety_checker=False, safety_checker=None,
)

pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
pipe.enable_model_cpu_offload()

pag_scales =  [0.0, 3.0]
guidance_scales = [0.0, 2.0]

grid = []
for pag_scale in pag_scales:
    for guidance_scale in guidance_scales:
        generator = torch.Generator(device="cpu").manual_seed(0)
        # generate image
        image = pipe(
            "a man with ray-ban sunglasses",
            negative_prompt="low quality, bad quality",
            num_inference_steps=20,
            generator=generator,
            eta=1.0,
            image=init_image,
            mask_image=mask_image,
            control_image=control_image,
            pag_scale=pag_scale,
            strength=0.99,
            guidance_scale=guidance_scale,
            controlnet_conditioning_scale=0.2
        ).images
        grid.append(image[0])

from diffusers.utils import make_image_grid
make_image_grid(grid, rows=len(pag_scales), cols=len(guidance_scales))
```

![pag_controlnet_inpaint](https://github.com/user-attachments/assets/9cd67619-8d16-4f4c-b80c-52c73fa3e1c2)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8875). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_inpaint.py:
```diff
@@ -0,0 +1,1550 @@
+                f"If image batch size is not 1, image batch size must be same as prompt batch size. image batch size: {image_batch_size}, prompt batch size: {prompt_batch_size}"
+            )
+
+    def prepare_control_image(
```
Could you add a `# Copied from` statement here from `pipeline_controlnet_inpaint.py`?

juancopi81 replied:
Sure!

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_inpaint.py:
```diff
@@ -0,0 +1,1550 @@
+        control_image = control_images if isinstance(control_image, list) else control_images[0]
+        controlnet_prompt_embeds = prompt_embeds
+
+        # 7.2 Create tensor stating which controlnets to keep
```
The numbering here onwards seems incorrect

juancopi81 replied:
Oh you're right! Changing it...

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_inpaint.py:
```diff
@@ -0,0 +1,1550 @@
+        mask_image,
+        height,
+        width,
+        callback_steps,
```
I think you can remove this parameter completely instead of passing None

juancopi81 replied:
Oh sure! I just removed it

a-r-r-o-w reviewed (approved):
Thanks for working on this! The PR looks absolutely perfect with all required PAG changes. I think you need to run `make style` for the CI to go green. @yiyixuxu WDYT?

juancopi81 commented:
Hi @a-r-r-o-w ,

Thank you very much for your feedback 🚀 ! I just updated my PR trying to address all the issues.

Thanks again.

juancopi81 commented:
Sorry I had a small mistake that made some tests failed. It should be solved now!

yiyixuxu commented on src/diffusers/pipelines/auto_pipeline.py:
```diff
@@ -944,7 +946,10 @@ def from_pretrained(cls, pretrained_model_or_path, **kwargs):
             enable_pag = kwargs.pop("enable_pag")
             if enable_pag:
-                orig_class_name = config["_class_name"].replace("Pipeline", "PAGPipeline")
+                if "controlnet" in kwargs:
```
can you provide an example where this code addition is needed? I think it should already be addressed in line 944 no?

juancopi81 replied:
Hi @yiyixuxu,

Thank you very much for your comments. Sorry if I am missing something here, but my idea is as follows:
There are 4 cases that need to be handled:

- No ControlNet and no Pag -> Keep **Pipeline**

- With ControlNet and no Pag -> Replace Pipeline with **ControlNetPipeline**

- No ControlNet and with Pag -> Replace Pipeline with **PAGPipeline**

- With ControlNet and with Pag -> Replace Pipeline with **ControlNetPAGPipeline**, which will override the first case.

That's why I added the nested conditions, but maybe I am missing something 🤔

yiyixuxu replied:
i see! thank for explaining it, you were absolute ly right! 


can we do this instead (to be consistent with AutoPipelineForText2Image and AutoPipelineForImage2Image)

```python
if "controlnet" in kwargs:
            orig_class_name = config["_class_name"].replace("Pipeline", "ControlNetPipeline")
        if "enable_pag" in kwargs:
            enable_pag = kwargs.pop("enable_pag")
            if enable_pag:
                orig_class_name = orig_class_name.replace("Pipeline", "PAGPipeline")
```
https://github.com/huggingface/diffusers/blob/73acebb8cfbd1d2954cabe1af4185f9994e61917/src/diffusers/pipelines/auto_pipeline.py#L363

juancopi81 replied:
Oh sure, done! This way is better!

yiyixuxu reviewed (commented):
thanks! super cool PR! I left one comment,
also, can we make it work with `from_pipe()`?

juancopi81 commented:
> thanks! super cool PR! I left one comment, also, can we make sure it work with `from_pipe()`?

Hi @yiyixuxu,

Thanks for your support and feedback. I updated my PR with a new test. I run it with:

`tests/pipelines/test_pipelines_auto.py::AutoPipelineFastTest::test_from_pipe_pag_controlnet_inpaint`

and it seems to be working fine. How do you like it? Would be enough for testing that it works with `from_pipe()`?

Please let me know 😃

yiyixuxu commented on tests/pipelines/test_pipelines_auto.py:
```diff
@@ -283,6 +283,26 @@ def test_from_pipe_pag_inpaint(self):
         pipe = AutoPipelineForInpainting.from_pipe(pipe_pag, enable_pag=False)
         assert pipe.__class__.__name__ == "StableDiffusionXLInpaintPipeline"
 
+    def test_from_pipe_pag_controlnet_inpaint(self):
```
nice! we can combine this test with `test_from_pipe_pag_inpaint` and make sure it covers all the scenarios similarly as in https://github.com/huggingface/diffusers/blob/main/tests/pipelines/test_pipelines_auto.py#L141

1. `test_from_pipe_pag_inpaint` test for the inpainting XL pipeline, we can update to test SD pipeline instead because we now have a controlnet + inpaint + PAG for SD but not for SDXL
2.  make sure :
     - we test use `from_pipe` on inpainting pipeline, inpainting controlnet pipeline and pag inpainting controlnet
     - for each, make sure we test the combination of `enable_pag` and controlnet` 

but if you want to skip the test, it's ok too! we can update the test when we have the sdxl pag inpainting pipeline as well

yiyixuxu reviewed (commented):
very nice!  I left some comments for the test, since all the from_pipe tests are based on SDXL, if you want to skip it, it is ok too! Just let us know!

Let's merge this soon

juancopi81 commented:
Thanks a lot @yiyixuxu for your feedback. I addressed your comment in [src/diffusers/pipelines/auto_pipeline.py](https://github.com/huggingface/diffusers/pull/8875/files/00f8c7a04d19275fcf81c4fc8b038bf8b3c02562#diff-edd3c8f30ab707862237a901e184533c4d44c6bf26bb307e965ef9dda28444f0), and remove for now the test I added so the PR can be merged. I can come back later to finish it properly or if someone want to take it, please let me know!

yiyixuxu reviewed (approved):
thanks so much for the PR
and very sorry I forgot about it and only merge it now

juancopi81 commented:
No worries @yiyixuxu! Thank you very much for the support. I am very happy that now is merged 🚀
